File: eio2024-ev/2024-12-08-ev/baas/solution/ahto_trie_prune.cpp
// optimaalne lahendus trie põhjal
// ehitab trie ainult sisendi esimese rea kohta
// teiste ridade kohta loeb ainult, millistele elementidele vastavad prefiksid neis ka esinevad
// teenib täispunktid

#include <iostream>
#include <string>
#include <map>
using namespace std;

struct Node {
	int row = 0;
	map<char, Node> next;
} root;

void add_word(int row, const string & word) {
	Node *curr = &root;
	for (auto c : word) {
		if (row > 1 and not curr->next.contains(c)) {
			return;
		}
		curr = &curr->next[c];
		if (curr->row == row - 1) {
			curr->row = row;
		}
	}
}

string best;

void find_best(const Node & curr = root, string word = "") {
	if (curr.row < root.row) {
		return;
	}
	if (best.length() < word.length()) {
		best = word;
	}
	for (const auto & [c, next] : curr.next) {
		find_best(next, word + c);
	}
}

int main() {
	int m, n;
	cin >> m >> n;
	for (int i = 0; i < n; ++i) {
		for (int j = 0; j < m; ++j) {
			string word;
			cin >> word;
			add_word(1 + i, word);
		}
	}
	root.row = n;
	find_best();
	cout << best << '\n';
}

File: eio2024-ev/2024-12-08-ev/baas/solution/ahto_trie_prune.py
# optimaalne lahendus trie põhjal
# ehitab trie ainult sisendi esimese rea kohta
# teiste ridade kohta loeb ainult, millistele elementidele vastavad prefiksid neis ka esinevad
# teenib täispunktid

m, n = (int(x) for x in input().split())

root = [n, {}]

def add_word(row, word):
	curr = root
	for c in word:
		if c not in curr[1]:
			if row > 1:
				return
			curr[1][c] = [0, {}]
		curr = curr[1][c]
		if curr[0] == row - 1:
			curr[0] = row

def print_trie(curr = root, char = '', level = 0):
	print(' ' * level, char, curr[0])
	for c in curr[1]:
		print_trie(curr[1][c], c, level + 1)

for row in range(n):
	words = input().split()
	for word in words:
		add_word(1 + row, word)
		#print_trie()

def find_best(curr = root, word = ''):
	global best
	if curr[0] < n:
		return
	if len(best) < len(word):
		best = word
	for c in sorted(curr[1]):
		find_best(curr[1][c], word + c)

best = ''
find_best()
print(best)
